From 57f0d968f1a766721a90df05dec746bfc3c41aa1 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 19 Sep 2015 01:38:29 -0400
Subject: [PATCH] Log potential duplicate entities

CB Bug caused us to get duplicated entities (same UUID).
This will auto clean them up.
---
 src/main/java/net/minecraft/server/WorldServer.java | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 49af0c4..6a6edf1 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -980,7 +980,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     protected void a(Entity entity) {
         super.a(entity);
         this.entitiesById.a(entity.getId(), entity);
-        this.entitiesByUUID.put(entity.getUniqueID(), entity);
+        final Entity old = this.entitiesByUUID.put(entity.getUniqueID(), entity); // EMC
+        if (old != null && old.getId() != entity.getId() && old.valid) {System.err.println("[DUPLICATE-REMOVAL] Removing duplicate " + old.getBukkitEntity()+"/"+old.getUniqueID() + " at " + old.getBukkitEntity().getLocation() + " is dupe of " + entity.getBukkitEntity() + "/" + entity.getUniqueID() + " at " + entity.getBukkitEntity().getLocation());} // EMC - auto remove duplicates
         Entity[] aentity = entity.aB();
 
         if (aentity != null) {
-- 
1.9.1

